packages/entity/src/ComposedEntityCacheAdapter.ts

```diff
@@ -40,11 +44,35 @@ export default class ComposedEntityCacheAdapter<TFields> extends EntityCacheAdap
           newUnfulfilledFieldValues.push(fieldValue);
         } else {
           retMap.set(fieldValue, cacheResult);
+          const fulfilledFieldValues = nullthrows(fulfilledFieldValuesByCacheIndex[i]);
+          fulfilledFieldValues.push(fieldValue);
         }
       }
       unfulfilledFieldValues = newUnfulfilledFieldValues;
     }
 
+    // Recache values from higher layers that were not found in lower layers
+    for (let i = 0; i < this.cacheAdapters.length; i++) {
+      const cacheAdapter = nullthrows(this.cacheAdapters[i]);
+      const hitsToCache = new Map<NonNullable<TFields[N]>, Readonly<TFields>>();
+      const missesToCache: NonNullable<TFields[N]>[] = [];
+      for (let j = i + 1; j < this.cacheAdapters.length; j++) {
```

Comment:
nit: add more inline comments about what this and the outer for loop index math is doing